src/Bicep.Core/Semantics/SemanticModel.cs

```diff
@@ -39,29 +40,26 @@ public class SemanticModel : ISemanticModel
         private readonly Lazy<ImmutableArray<DeclaredResourceMetadata>> declaredResourcesLazy;
         private readonly Lazy<ImmutableArray<IDiagnostic>> allDiagnostics;
 
-        public SemanticModel(Compilation compilation, BicepSourceFile sourceFile, IFileResolver fileResolver, IBicepAnalyzer linterAnalyzer, RootConfiguration configuration, IFeatureProvider features, IApiVersionProvider apiVersionProvider)
+        public SemanticModel(Compilation compilation, BicepSourceFile sourceFile, IFileResolver fileResolver, IBicepAnalyzer linterAnalyzer, RootConfiguration configuration, IFeatureProvider features)
         {
             Trace.WriteLine($"Building semantic model for {sourceFile.FileUri} ({sourceFile.FileKind})");
 
-            Compilation = compilation;
-            SourceFile = sourceFile;
-            Configuration = configuration;
-            Features = features;
-            ApiVersionProvider = apiVersionProvider;
-            FileResolver = fileResolver;
+            this.Compilation = compilation;
+            this.SourceFile = sourceFile;
+            this.Configuration = configuration;
+            this.Features = features;
+            this.FileResolver = fileResolver;
 
             // create this in locked mode by default
             // this blocks accidental type or binding queries until binding is done
             // (if a type check is done too early, unbound symbol references would cause incorrect type check results)
-            var symbolContext = new SymbolContext(compilation, this);
-            SymbolContext = symbolContext;
+            this.SymbolContext = new SymbolContext(compilation, this);
+            this.Binder = new Binder(compilation.NamespaceProvider, features, sourceFile, this.SymbolContext);
+            this.ApiVersionProvider = new ApiVersionProvider(features, this.Binder.NamespaceResolver.GetAvailableResourceTypes());
```

Comment:
Here's the code where I'm concerned about perf...

Reply:
@StephenWeatherford, totally agree with you about making sure we are meeting performance needs, can you please elaborate what is the action you propose we should take to assuage these concerns?

Reply:
I ran:

## Results Summary

```bash
// Benchmark Process 47708 has exited with code 0.

Mean = 2.483 s, StdErr = 0.024 s (0.99%), N = 98, StdDev = 0.242 s
Min = 2.167 s, Q1 = 2.303 s, Median = 2.403 s, Q3 = 2.635 s, Max = 3.131 s
IQR = 0.332 s, LowerFence = 1.805 s, UpperFence = 3.133 s
ConfidenceInterval = [2.400 s; 2.566 s] (CI 99.9%), Margin = 0.083 s (3.35% of Mean)
Skewness = 1.01, Kurtosis = 3.02, MValue = 2.51

// ** Remained 0 (0.0%) benchmark(s) to run. Estimated finish 2023-05-22 9:11 (0h 0m from now) **
Successfully reverted power plan (GUID: 8c5e7fda-e8bf-4a96-9a85-a6e23a8c635c FriendlyName: High performance)
// ***** BenchmarkRunner: Finish  *****

// * Export *
  BenchmarkDotNet.Artifacts\results\Bicep.Tools.Benchmark.Compilation-report.csv
  BenchmarkDotNet.Artifacts\results\Bicep.Tools.Benchmark.Compilation-report-github.md
  BenchmarkDotNet.Artifacts\results\Bicep.Tools.Benchmark.Compilation-report.html

// * Detailed results *
Compilation.'Get compilation diagnostics for each dataset': DefaultJob
Runtime = .NET 7.0.5 (7.0.523.17405), X64 RyuJIT AVX2; GC = Concurrent Workstation
Mean = 2.483 s, StdErr = 0.024 s (0.99%), N = 98, StdDev = 0.242 s
Min = 2.167 s, Q1 = 2.303 s, Median = 2.403 s, Q3 = 2.635 s, Max = 3.131 s
IQR = 0.332 s, LowerFence = 1.805 s, UpperFence = 3.133 s
ConfidenceInterval = [2.400 s; 2.566 s] (CI 99.9%), Margin = 0.083 s (3.35% of Mean)
Skewness = 1.01, Kurtosis = 3.02, MValue = 2.51
-------------------- Histogram --------------------
[2.098 s ; 2.255 s) | @@@@@@@@@
[2.255 s ; 2.393 s) | @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
[2.393 s ; 2.535 s) | @@@@@@@@@@@@@@@@@@@@@
[2.535 s ; 2.590 s) | @@
[2.590 s ; 2.728 s) | @@@@@@@@@@@@
[2.728 s ; 2.919 s) | @@@@@@@
[2.919 s ; 3.091 s) | @@@@@@@
[3.091 s ; 3.200 s) | @
---------------------------------------------------


// * Legends *
  Mean      : Arithmetic mean of all measurements
  Error     : Half of 99.9% confidence interval
  StdDev    : Standard deviation of all measurements
  Median    : Value separating the higher half of all measurements (50th percentile)
  Gen0      : GC Generation 0 collects per 1000 operations
  Gen1      : GC Generation 1 collects per 1000 operations
  Allocated : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 s       : 1 Second (1 sec)

// * Diagnostic Output - EventPipeProfiler *
Exported 1 trace file(s). Example:
C:\Users\asilverman\src\ms\bicep\BenchmarkDotNet.Artifacts\Bicep.Tools.Benchmark.Compilation.Get_compilation_diagnostics-20230522-085918.speedscope.json

// * Diagnostic Output - MemoryDiagnoser *


// ***** BenchmarkRunner: End *
```

Full contents of the file:
[Bicep.Tools.Benchmark.Compilation.Get_compilation_diagnostics-20230522-085918.speedscope.json.gz](https://github.com/Azure/bicep/files/11533570/Bicep.Tools.Benchmark.Compilation.Get_compilation_diagnostics-20230522-085918.speedscope.json.gz)

Reply:
@StephenWeatherford, one more thought here is that this feature is experimental, its performance can be tuned in a separate PR. I would like to be able to work on the problem of performance tuning separate from the functional viability so that it's easier to review these PRs and so that we can have a clear decoupling of the code contributions.

Some ideas that came up during our conversations are:
- Is `ApiVersionProvider` really needed? Could we fold this behavior into the `AzResourceTypeProvider`?
- Do we need to re-generate the dictionary that `ApiVersionProvider` stores in its own state or could it be stored in some other objects state?
- Can we expose the <scope,apiversions> dictionary as a method of the AzResourceTypeProvider?

Reply:
Hadn't seen this yet.  As discussed, as long as the current mainline behavior doesn't regress (much)....  Thx

Reply:
Apologies, I forgot that the use-recent-api-versions linter rule has to be turned on for the code in question to run.

I did some quick perf testing on my MacBook Pro (2.6 GHz 6-Core Intel Core i7), and am getting an average of about 40ms each time the code is run.  Note that the performance shouldn't be affected by file size, only by the number of files being recompiled.  So for each character typed in normal circumstances I would expect an extra 40ms delay.  If you make changes that cause multiple files to be recompiled, it would be 40ms per file.

That's not ideal but doesn't seem likely to be noticed under normal circumstances (unless you have a much slower computer).  I'm not seeing the difference while typing on my Mac.  Ideally we'll do some perf tuning of all the linter rules at some point.  Until then, I don't feel like we need to block the merge.